File: backend/models/running_session.py
"""
Running Session database model
"""
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Integer, Float, DateTime, ForeignKey, JSON, Text
from sqlalchemy.dialects.mysql import CHAR
from sqlalchemy.orm import relationship
from database.database import Base


class RunningSession(Base):
    __tablename__ = "running_sessions"

    id = Column(CHAR(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(CHAR(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    date = Column(DateTime, nullable=False, index=True)
    distance_km = Column(Float, nullable=False)
    avg_pace = Column(Float, nullable=True)  # minutes per km
    max_speed = Column(Float, nullable=True)  # km/h
    duration_seconds = Column(Integer, nullable=False)
    route_geojson = Column(JSON, nullable=True)  # GeoJSON format
    calories_burned = Column(Integer, nullable=True)
    elevation_gain = Column(Float, nullable=True)  # meters
    avg_heart_rate = Column(Integer, nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="running_sessions")

File: lib/models/running_session.dart
"""
RunningSession model mirroring backend schema
"""
class RunningSession {
  final String id;
  final String userId;
  final DateTime date;
  final double distanceKm;
  final double? avgPace;
  final double? maxSpeed;
  final int durationSeconds;
  final int? caloriesBurned;
  final double? elevationGain;
  final int? avgHeartRate;
  final String? notes;

  const RunningSession({
    required this.id,
    required this.userId,
    required this.date,
    required this.distanceKm,
    this.avgPace,
    this.maxSpeed,
    required this.durationSeconds,
    this.caloriesBurned,
    this.elevationGain,
    this.avgHeartRate,
    this.notes,
  });

  factory RunningSession.fromJson(Map<String, dynamic> j) => RunningSession(
    id: j['id'] as String,
    userId: j['user_id'] as String,
    date: DateTime.parse(j['date'] as String),
    distanceKm: (j['distance_km'] as num).toDouble(),
    avgPace: (j['avg_pace'] as num?)?.toDouble(),
    maxSpeed: (j['max_speed'] as num?)?.toDouble(),
    durationSeconds: j['duration_seconds'] as int,
    caloriesBurned: j['calories_burned'] as int?,
    elevationGain: (j['elevation_gain'] as num?)?.toDouble(),
    avgHeartRate: j['avg_heart_rate'] as int?,
    notes: j['notes'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'date': date.toIso8601String(),
    'distance_km': distanceKm,
    'duration_seconds': durationSeconds,
    'avg_pace': avgPace,
    'max_speed': maxSpeed,
    'calories_burned': caloriesBurned,
    'elevation_gain': elevationGain,
    'avg_heart_rate': avgHeartRate,
    'notes': notes,
  };
}
